packages/dds/tree/src/feature-libraries/sequence-change-family/rebase.ts

```diff
@@ -0,0 +1,116 @@
+} from "../../changeset";
+import { clone, fail } from "../../util";
+import { SequenceChangeset } from "./sequenceChangeset";
+
+/**
+ * Rebases `change` over `base` assuming they both apply to the same initial state.
+ * @param change - The changeset to rebase.
+ * @param base - The changeset to rebase over.
+ * @returns A changeset that performs the changes in `change` but does so assuming `base` has been applied first.
+ *
+ * WARNING! This implementation is incomplete:
+ * - Marks that affect existing content are removed instead of muted when rebased over the deletion of that content.
+ *   This prevents us from then reinstating the mark when rebasing over the revive.
+ * - Tombs are not added when rebasing an insert over a gap that is immediately left of deleted content.
+ *   This prevents us from being able to accurately track the position of the insert.
+ * - Tiebreak ordering is not respected.
+ * - Support for moves is not implemented.
+ * - Support for slices is not implemented.
+ */
+export function rebase(change: SequenceChangeset, base: SequenceChangeset): SequenceChangeset {
+    const fields = rebaseFieldMarks(change.marks, base.marks);
+    return {
+        marks: fields,
+    };
+}
+
+function rebaseFieldMarks(change: T.FieldMarks, base: T.FieldMarks): T.FieldMarks {
+    const fields: T.FieldMarks = {};
+    for (const key of Object.keys(change)) {
+        if (key in base) {
+            fields[key] = rebaseMarkList(change[key], base[key]);
+        } else {
+            fields[key] = clone(change[key]);
+        }
+    }
+    return fields;
+}
+
+function rebaseMarkList(currMarkList: T.MarkList, baseMarkList: T.MarkList): T.MarkList {
+    const factory = new MarkListFactory();
+    let iBase = 0;
+    let iCurr = 0;
+    let nextCurrMark: T.Mark | undefined = currMarkList[iCurr];
+    let nextBaseMark: T.Mark | undefined = baseMarkList[iBase];
+    while (nextCurrMark !== undefined && nextBaseMark !== undefined) {
+        let currMark: T.Mark = nextCurrMark;
+        let baseMark: T.Mark = nextBaseMark;
+        nextCurrMark = undefined;
+        nextBaseMark = undefined;
+
+        if (isAttachGroup(currMark) || isReattach(currMark)) {
+            // TODO: respect tiebreak
+            factory.pushContent(clone(currMark));
+            nextBaseMark = baseMark;
+        } else if (isAttachGroup(baseMark) || isReattach(baseMark)) {
+            factory.pushOffset(getMarkLength(baseMark));
+            nextCurrMark = currMark;
+        } else {
+            const currMarkLength = getMarkLength(currMark);
+            const baseMarkLength = getMarkLength(baseMark);
+            if (currMarkLength < baseMarkLength) {
+                [baseMark, nextBaseMark] = splitMark(baseMark, currMarkLength);
+            } else if (currMarkLength > baseMarkLength) {
+                [currMark, nextCurrMark] = splitMark(currMark, baseMarkLength);
+            }
+            const rebasedMark = rebaseMark(currMark, baseMark);
```

Comment:
Maybe this is a simpler algorithm than the alternatives, but it seems a bit strange that we may often split the mark we are rebasing, only to immediately have the `MarkListFactory` combine it back together.

Reply:
I agree it's not ideal. If we invest in a nice iterator we may be able to avoid that. Having this simpler version checked in (and hardened as we find issues with it) will make it safer to re-write it later if needed.